### PR TITLE
feat!:(2.0 global services)

### PR DESCRIPTION
**Pull request by jacoobes**
# sern 2.0!!!
- Test it at your own risk, this PR is in active development. If you decide to test and find some bugs, make sure to report it in our server

# INCLUDES BREAKING CHANGES!

### CODE REVIEW

**Comment by EvolutionX-10** [comment]
review is in progress, will do it some files at a time 😅

File: src/handler/events/messageHandler.ts [hunk 1]
```diff
@@ -5,48 +5,45 @@ import type { Message } from 'discord.js';
...
 import Context from '../structures/context';
-import * as Files from '../utilities/readFile';
-import type { TextCommand } from '../structures/module';
 import { CommandType, PayloadType } from '../structures/enums';
-import { asyncResolveArray } from '../utilities/asyncResolveArray';
+import { arrAsync } from '../utilities/arrAsync';
 import { controller } from '../sern';
+import type { CommandModule, TextCommand } from '../../types/module';
+import { handleError } from '../contracts/errorHandling';
+import type { ModuleStore } from '../structures/moduleStore';
```
Comment (EvolutionX-10) [inline src/handler/events/messageHandler.ts:14]: we gotta pretty the imports man

File: src/handler/structures/enums.ts [hunk 2]
```diff
@@ -16,38 +17,46 @@ export enum CommandType {
     /**
      * The CommandType for text commands
      */
-    Text = 0b00000000001,
+    Text = 1,
```
Comment (EvolutionX-10) [inline src/handler/structures/enums.ts:20]: finally good numbers!

File: src/handler/utilities/functions.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+ * Used for singleton in iti
+ * @param value
+ */
+export const _const = <T>(value: T) => () => value;
+/**
+ * A function that returns another function
+ * Used for transient in iti
+ * @param value
+ */
+export const transient = <T>( value : T) => () => _const(value);
```
Comment (EvolutionX-10) [inline src/handler/utilities/functions.ts:16]: thats coool

**Review by EvolutionX-10 — APPROVED** [review]
why test/esm/src/index.ts is empty file?

**Comment by jacoobes** [comment]
Was gonna add testing but I'm lazy

File: src/handler/structures/enums.ts (hunk 2 above)
Comment (jacoobes) [reply src/handler/structures/enums.ts:20]: true

File: src/handler/events/messageHandler.ts (hunk 1 above)
Comment (jacoobes) [reply src/handler/events/messageHandler.ts:14]: one day

File: .eslintrc [hunk 4]
```diff
@@ -6,6 +6,8 @@
         "@typescript-eslint/no-non-null-assertion": "off",
         "quotes": [2, "single", { "avoidEscape": true, "allowTemplateLiterals": true }],
         "semi": ["error", "always"],
-        "@typescript-eslint/no-empty-interface": 0
+        "@typescript-eslint/no-empty-interface": 0,
+        "@typescript-eslint/ban-types" : 0,
+        "@typescript-eslint/no-explicit-any": "off"
```
[the remark this answers is not in the document]
Comment (jacoobes) [reply .eslintrc:11]: xD

File: src/handler/structures/errors.ts [hunk 5]
```diff
@@ -31,4 +31,8 @@ export enum SernError {
      * Unsupported feature attempted to access at this time
      */
     NotSupportedYet = `This feature is not supported yet`,
+    /**
+     * Required Dependency not found
+     */
+    MissingRequired = `@sern/client is required, was not found`
```
Comment (Murtatrxx) [inline src/handler/structures/errors.ts:37]: Adding "but" would help

**Review by Murtatrxx — COMMENTED** [review]
Looks good overall, but needs small improvements

File: src/handler/dependencies/provider.ts [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+        keys.map(key => Result.wrap(() => container.get(key)).unwrapOr(undefined)) as MapDeps<T, V>;
+}
+
+/**
+ * Returns the underlying data structure holding all dependencies.
+ * Exposes some methods from iti
+ */
+export function useContainerRaw() {
+    return containerSubject.getValue();
+}
```
[the remark this answers is not in the document]
Comment (jacoobes) [reply src/handler/dependencies/provider.ts:83]: understandable

[the remark this answers is not in the document]
Comment (jacoobes) [reply src/handler/dependencies/provider.ts:83]: will refactor in future possibly

**Pull request merged**